### PR TITLE
fix(rl): cap policy-gradient scalar normalization

### LINKED ISSUE #1319: P1: Reduce RL policy-gradient normalization after near-zero trusted-gradient signal
## Problem
Latest RL policy online advantage ledger reports the true-gradient computation path is now active, but the signal is flattened by the current gradient normalization scale:

- Policy advantage report: `runtime-artifacts/rl-control-loop/20260522T032832Z-policy-advantage.json`
- Ledger finding: candidate mode is `construction-priority.pg.{territory,resource,risk-aware}-seed.v1` / shadow; online utility remains `MIXED` for the 65th consecutive report.
- Regressions called out by the ledger:
  - `P1 — Gradient near-zero due to 1e9 normalization — FIX: Reduce normalization to 1e3-1e4`
  - `P1 — trustedGradientUpdate still false — gate correctly blocks, normalization fix should resolve`
- Training ledger: `runtime-artifacts/rl-control-loop/20260522T035208Z-training-ledger.json` reports `trainingDidRun=true`, 31 candidate evaluation iterations, 25 policy update iterations, and active gradient trust building.
- Current main after #1317: `b074628b`; an active Tencent validation run `tencent-pg-20260522t034423z` is still in progress, so additional paid compute should stay guarded, but code correction can proceed offline.

## Scope
Implement the smallest code/test change needed to make the RL policy-gradient update preserve informative gradient magnitude:

1. Replace or parameterize the hard-coded `1e9` gradient normalization factor with a scale in the `1e3`–`1e4` range, using repository naming/config conventions.
2. Add/update tests proving small but meaningful policy-gradient deltas are not flattened to near-zero while safety/trust gates still block genuinely untrusted updates.
3. Preserve existing scorecard, reward, and safety semantics; no official MMO writes and no paid compute launch from this issue.

## Acceptance
- Controller verification passes for the touched scripts/tests.
- A post-merge Loop A/B or guarded Tencent run shows non-near-zero gradient signal and `trustedGradientUpdate` can turn true when the trust criteria are met.
- Update #879/#1032/#1233/#1299 evidence after PR creation and after validation.

Refs #879 #1032 #1233 #1299

### PR DESCRIPTION
**Pull request by lanyusea**
## Summary
- Caps the policy-gradient scalar reward-weight normalization denominator so lower reward tiers remain observable after near-zero trusted-gradient evidence.
- Adds explicit evidence fields for source max weight and normalization cap in policy-update scalar reward metadata.
- Updates RL training-runner tests to verify nonzero territory/combat signal after the cap.

## Verification
- `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
- `python3 -m unittest scripts.test_screeps_rl_training_runner -q` (95 tests)
- `git diff --check`

Fixes #1319

## Safety
- Offline RL training-runner change only; no official MMO runtime upload or production Screeps tick-loop behavior is changed directly.

### CODE REVIEW

**Comment by lanyusea** [comment]
Review-fix checkpoint (controller-verified):

- Codex triage artifact: `/tmp/pr1320-codex-triage.md`.
- Classifications:
  - `PRRT_kwDOSMSsO86EAMhc` — `FIX`: cap denominator was changing scalar update magnitude; fix preserves source-scale update magnitude while keeping cap-normalized audit evidence.
  - `PRRT_kwDOSMSsO86EAOpl` — `FIX`: exported `sourceMaxComponentWeight`, `normalizationCap`, and `scalarRewardScaleFactor` in `gradientEstimation` for downstream auditability.
- Commit pushed: `4addda1c` (`fix(rl): preserve capped gradient update scale`), authored by `lanyusea's bot <lanyusea@gmail.com>`.
- Controller verification:
  - `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
  - `python3 -m unittest scripts.test_screeps_rl_training_runner -q` — 95 tests OK
  - `git diff --check`

Gate now pending exact-head `prod-ci`, CodeRabbit re-review, and fresh GraphQL review-thread query on `4addda1c` before resolving threads/merge.

**Comment by lanyusea** [comment]
PR #1320 follow-up review-fix checkpoint (2026-05-22T05:37Z)

Exact head now: `17c6274183b2414afcfbf2c946edaa5a0926ff2a` (`fix(rl): apply scalar normalization cap to updates`).

What changed after the latest CodeRabbit thread:
- `PRRT_kwDOSMSsO86EAZxQ` was classified `FIX` / critical by Codex triage because the previous compensation made the `10_000` cap affect evidence but not the applied update gradient.
- Commit `17c62741` now applies the cap-normalized scalar reward signal directly to the update gradient / contribution evidence while keeping `sourceMaxComponentWeight`, `normalizationFactor`, and full-precision `scalarRewardScaleFactor` for auditability.
- The earlier rounding thread `PRRT_kwDOSMSsO86EAXCQ` remains addressed by retaining full-precision `scalarRewardScaleFactor`.
- New minor CodeRabbit coverage request `PRRT_kwDOSMSsO86EAhL8` was Codex-triaged as `ADVISORY_ONLY`: it asks for optional below-cap test coverage and does not identify a runtime/safety/deployment defect; quick local evidence for weights 1000/100/10/1 returned `normalizationFactor=1000` and `scalarRewardScaleFactor=1.0` under the current contract.

Controller verification on the final tree before commit:
- `git diff --check`
- `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
- `python3 -m unittest scripts.test_screeps_rl_training_runner -q` (97 tests OK)

Current gate:
- `prod-ci` is green on exact head `17c62741`.
- CodeRabbit is still `PENDING` on exact head, so I am not resolving the remaining outdated/advisory threads or merging yet. Next scheduler pass should re-query CodeRabbit + GraphQL review threads and then resolve stale/advisory/fixed threads if the bot has finished without new critical findings.

**Comment by lanyusea** [comment]
PR #1320 latest review-fix checkpoint (2026-05-22T05:50Z)

Exact head is now `2c1bd77a6bd5e55cbecb593586cd07e579a47c7c` (`fix(rl): preserve unrounded scalar update gradients`).

Additional triage/fix after CodeRabbit completed the prior cycle:
- `PRRT_kwDOSMSsO86EAmMI` was triaged as `OWNER_DECISION` / non-critical by Codex: reverting to pre-cap/source-max gradients would undo #1319's normalization-cap invariant, so the cap-normalized update behavior remains.
- `PRRT_kwDOSMSsO86EAoi_` was triaged as `FIX` / critical: rounded display evidence was feeding back into actual update math. Commit `2c1bd77a` now keeps unrounded scalar gradients/EMA values for update math while preserving rounded evidence fields.
- Existing minor below-cap coverage thread remains `ADVISORY_ONLY` from Codex triage.

Controller verification on the final tree before commit:
- `git diff --check`
- `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`
- `python3 -m unittest scripts.test_screeps_rl_training_runner -q` (98 tests OK)

Current gate after push:
- exact-head `prod-ci` started at 2026-05-22T05:50:19Z and is in progress.
- CodeRabbit is pending on exact head `2c1bd77a`.
- Next scheduler pass should re-query exact-head checks and GraphQL review threads; resolve fixed/outdated/advisory threads only after CodeRabbit finishes and no new critical findings remain.

**Comment by lanyusea** [comment]
Scheduler checkpoint dispatched Codex review-fix/triage for PR #1320.

- Worktree: `/root/screeps-worktrees/issue-1319-rl-gradient-normalization`
- Background session: `proc_6f3b8115f855` (OS pid `2109217`)
- Starting head: `2c1bd77a6bd5e55cbecb593586cd07e579a47c7c`
- Review-thread evidence: 6 unresolved total / 5 active non-outdated from `/tmp/screeps-scheduler-audit/pr-review-threads.json`
- Codex prompt requires per-thread classifications, `/tmp/pr1320-codex-triage.md`, `git diff --check`, `py_compile`, and `python3 -m unittest scripts.test_screeps_rl_training_runner -q`, with Codex-authored commit if a critical fix is required.

No push/merge will be attempted until the process exits, controller verification passes, exact-head checks rerun, and review threads are re-queried.

**Comment by lanyusea** [comment]
PR #1320 review-fix update (2026-05-22T06:21Z)

- New Codex-authored commit pushed: `d6fe4e7f3ad93b52ca3fa1b30d7e94670cfbc07d` — `fix(rl): preserve raw EMA gradient state`
- Commit files: `scripts/screeps_rl_training_runner.py`, `scripts/test_screeps_rl_training_runner.py`
- Controller verification before commit: `git diff --check`; `python3 -m py_compile scripts/screeps_rl_training_runner.py scripts/test_screeps_rl_training_runner.py`; `python3 -m unittest scripts.test_screeps_rl_training_runner -q` → 99 tests OK
- Codex triage file: `/tmp/pr1320-codex-triage.md`

Triage summary:
- `PRRT_kwDOSMSsO86EA2h3`: FIX — raw EMA state was not round-trippable; patched by exporting rounded `emaGradient`, preserving full-precision `rawEmaGradient`, and consuming raw EMA for update math.
- `PRRT_kwDOSMSsO86EAhL8`: ADVISORY_ONLY — below-cap edge-case coverage is useful but not critical; no patch.
- `PRRT_kwDOSMSsO86EAmMI`: OWNER_DECISION / non-critical conflict with #1319 invariant; no revert to pre-cap scaling.
- `PRRT_kwDOSMSsO86EA2h5`: ADVISORY_ONLY — assertion/style coverage request, no critical defect after raw EMA fix.
- `PRRT_kwDOSMSsO86EA2h9`: ADVISORY_ONLY — metadata assertion request, no critical defect.
- `PRRT_kwDOSMSsO86EAXCQ`: RESOLVE_STALE_OR_OUTDATED — original scalarRewardScaleFactor zeroing issue is already fixed on current head.

Current exact-head gate after push:
- PR head: `d6fe4e7f3ad93b52ca3fa1b30d7e94670cfbc07d`
- `prod-ci` is IN_PROGRESS.
- `CodeRabbit` is PENDING.
- GraphQL review threads after push: 5 unresolved total / 4 active non-outdated; do not resolve until CodeRabbit finishes and threads are re-queried on the exact head.

Next action: re-query exact-head checks and review threads after CodeRabbit/prod-ci complete; resolve fixed/advisory/stale threads with the triage evidence if no new critical findings appear, then run the merge gate.

**Comment by lanyusea** [comment]
PR #1320 latest-thread triage complete (2026-05-22T06:28Z)

- Codex latest-thread triage process: `proc_b17125e211c4` exited cleanly with no repo changes needed.
- Triage file: `/tmp/pr1320-codex-triage-latest.md`
- Worktree status: clean at head `d6fe4e7f3ad93b52ca3fa1b30d7e94670cfbc07d`.
- Exact-head checks now pass: `prod-ci` SUCCESS and `CodeRabbit` SUCCESS.

Latest triage result:
- `PRRT_kwDOSMSsO86EAXCQ`: `RESOLVE_STALE_OR_OUTDATED` — scalarRewardScaleFactor zeroing allegation is fixed on current head.
- `PRRT_kwDOSMSsO86EAhL8`: `ADVISORY_ONLY` — below-cap edge-case coverage request is non-critical.
- `PRRT_kwDOSMSsO86EAmMI`: `OWNER_DECISION`/non-critical duplicate of the scale-invariance policy conflict; no code change because #1319 intentionally uses cap-normalized update magnitude as an offline/shadow gated fix.
- `PRRT_kwDOSMSsO86EA2h9`: `ADVISORY_ONLY` — metadata assertion request is non-critical.
- `PRRT_kwDOSMSsO86EBK8M`: `OWNER_DECISION`/duplicate of `EAmMI`; no code change.

Controller is resolving the triaged review threads and will run the merge gate only after a fresh GraphQL/PR/check/Project re-query.
